qiskit/pulse/ir/ir.py

```diff
@@ -279,6 +274,35 @@ def edge_map(_x, _y, _node):
 
         return block
 
+    def copy(self) -> SequenceIR:
+        """Create a copy of ``SequenceIR``.
+
+        The returned copy can be safely mutated without affecting the original object, while immutable
+        objects are still passed as reference for memory efficiency.
+
+        .. warning::
+            If node data (not a nester IR object) is mutated via ``.sequence`` or ``.elements`` it will
+            affect the original IR object.
+
+        ``SequenceIR`` is poorly suited for both shallow and deep copy. A shallow copy
+        will contain references to mutable properties like ``sequence`` and ``time_table``.
+        A deep copy on the other hand will needlessly copy immutable objects like
+        :class:`.qiskit.pulse.Instruction`. This function returns a "semi-deep" copy -
+        A new object containing new objects for ``sequence`` and ``time_table``. However,
+        node data of type :class:`.qiskit.pulse.Instruction` will be passed as a reference.
+        Nested ``SequenceIR`` objects are copied using the same logic.
+
+        Returns: A copy of the object.
+        """
+        copied = self.__class__(self.alignment)
```

Comment:
This is not really significant, but you can save the cost to instantiate empty (unused) dict and PyDAG by controlling instance creation with `__new__`.
```python
copied = object.__new__(self.__class__)
copied._time_table = ...
```
Since empty dict and PyDAG are lightweight this is probably just a matter of preference.